### ISSUE TITLE
UnifiedMultimodalNetwork: weight-streaming WeightRegistry collision between original and clone (MoreData_ShouldNotDegrade)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1738]
## Summary
`UnifiedMultimodalNetworkTests.MoreData_ShouldNotDegrade` fails (also **in isolation**) with:
```
System.InvalidOperationException : WeightRegistry.Configure: existing streaming pool has 22 registered entries.
Unregister all weights first, or call Reset() to forcibly drop them.
```
Split out of #1623 (one of the two model bugs still live there alongside #1737).

## Root cause (verified)
`MoreData_ShouldNotDegrade` builds `network1` **and** `network2 = network1.Clone()`. UnifiedMultimodalNetwork's structural parameter estimate crosses the `DefaultStreamingThresholdParams` (500M) gate, so each instance auto-enables weight streaming on first `Train` and registers its trainable tensors with the **process-global, single-tenant** `WeightRegistry`. `network1` registers (22 entries); then `network2.Train` → `TryAutoEnableWeightStreaming` → `ConfigureWeightLifetime` → `WeightRegistry.Configure`, which hard-throws because the pool is already populated (the clone also shares `network1`'s weight tensors via copy-on-write, so they're the same registered objects).

This is the same single-tenant-registry class as the Phi3Vision/SmolVLM streaming-reset work in #1722 — but it occurs **within one test** (original + clone), so the between-tests `ResetsWeightStreamingBetweenTests` hook does not help.

## Fix options (need a decision)
1. **Don't auto-stream a model that fits in memory.** UnifiedMultimodal at its default config (~500-600M) is just over the conservative 500M threshold but does not actually need streaming on a normal host. Gating auto-stream on real memory pressure (not pure param count) would avoid the collision — but touches the shared streaming trigger.
2. **Multi-tenant / takeover-safe `WeightRegistry`** (AiDotNet.Tensors) so a second foundation model can configure without throwing. A blind `Reset()` on collision is NOT safe — it can drop the first model's paged weights mid-life.
3. **Idempotent registration for shared (CoW-cloned) tensors** so a clone doesn't re-register tensors already in the pool.

## Acceptance
- `UnifiedMultimodalNetworkTests.MoreData_ShouldNotDegrade` green in isolation and in-shard; fix is production-ready (no blind global reset that risks corrupting a live model).

Refs #1623.

**Comment by ooples** [comment]
## Performance pass — profiling confirms the HeavyTimeout deferral is inherent, not a quick fix

Profiled `UnifiedMultimodalNetwork<float>()` training under the exact ModelFamily test conditions (CpuEngine + single-threaded determinism BLAS, `OPENBLAS_NUM_THREADS=1`), via a tight training driver + dotnet-trace thread-time sampling:

| Metric | Value |
|---|---|
| ParameterCount | **539,897,140** (~540M, just over the 500M streaming gate) |
| Inference forward (`Predict`) | ~14 ms/iter (fast — uses the compiled/accelerated path) |
| Full train step (fwd+backward+AdamW) | **~900–1090 ms/iter** |
| Managed allocation | **~2.3 GB/iter**, driving repeated gen2 GCs |
| Extrapolated `MoreData` (250 steps) | **~225–270 s** (gate = 120 s) |
| Thread-time sampling | **96% UNMANAGED_CODE_TIME** (native GEMM + memory) |

**Where the time goes:** the training step (not inference) is the cost. The ~2.3 GB/iter is the backward pass's gradient tensors (~2.16 GB for 540M fp32 params), and the 96%-unmanaged time is single-threaded BLAS GEMM in the backward + the matching native memory traffic.

**What is NOT the bottleneck:** I tried rewriting the dense `AdamW.Step` in-place (it allocates ~13 param-size temporaries per param per step). It produced **no measurable improvement** — those temporaries are pooled (not GC-allocated), so they were never the dominant cost. Reverted (no benefit, shared-optimizer risk).

**Conclusion:** the slowness is the single-threaded backward GEMM + gradient allocation of a genuine 540M-param model. Under the test's determinism pin (`SetDeterministicMode` → 1 BLAS thread) this is irreducible without deep backward/tape work — gradient-buffer pooling / optimizer-in-backward / backward buffer-reuse — which is the **#1662** epic, not a localized model fix. Shrinking the model is off the table (never-shrink-tests). So the `HeavyTimeout` tag (PR #1739) is the correct, evidence-backed resolution: deferred to the nightly lane, graduates back when #1662's backward optimizations land.

**Comment by ooples** [comment]
## Correction — the slowness is NOT inherent. The fused fast path is broken for this model.

My earlier "irreducible without deep tape work" comment was wrong. Deeper profiling found the real cause:

**UnifiedMultimodal silently falls off the compiled fused-training fast path and runs the eager autograd tape every step — ~17× slower.**

`NeuralNetworkBase.TryTrainWithFusedOptimizer` → `CompiledTapeTrainingStep.TryStepWithFusedOptimizer` throws a **`NullReferenceException`** inside the Tensors package's `CompiledTrainingPlan.ConfigureOptimizerFloat` step closure (`StepEager` → `b__0`), and is caught + falls back to the eager tape. The eager path is the one that allocates ~2.3 GB/iter and runs at ~1 s/step (≈250 s for the 250-step `MoreData`).

Evidence:
- The fused path is the **industry-standard** approach (forward + backward + fused optimizer in one compiled SIMD kernel, **zero materialized gradient tensors** — the codebase's own comment compares it to PyTorch's caching allocator).
- A **simple FeedForwardNeuralNetwork with default AdamW uses the fused path cleanly** (no fallback) — so the fast path works; the NRE is specific to UnifiedMultimodal's structure (4 modality encoders, only the text path exercised by a single `[768]` input).
- The eager `AdamW.Step` tolerates params with no gradient (`continue`); the fused plan does not handle UnifiedMultimodal's case and NREs.

**Real fix (in AiDotNet.Tensors):** make `CompiledTrainingPlan.ConfigureOptimizerFloat` / its step closure robust to whatever UnifiedMultimodal's graph hits (most likely a param that receives no gradient, or a shape edge case), then release + bump the package. Once the fused path engages, UnifiedMultimodal should train at fused speed and likely **graduate out of HeavyTimeout** — no model shrink, no eager-tape rewrite needed.

Pinpointing the exact null requires debugging the Tensors source at the matching released version (0.104.6) with symbols + a ProjectReference repro — tracked as the next step. The HeavyTimeout tag from PR #1739 stays in the meantime (gate stays green), but it's now a *known, fixable* deferral with a concrete root cause, not "inherent."

**Comment by ooples** [comment]
## Root cause pinned + fixed (AiDotNet.Tensors PR #712)

Debugged against the Tensors v0.104.6 source (ProjectReferenced, symbols on). The fused training fast path NRE's at `CompiledTrainingPlan.ConfigureOptimizerFloat`'s per-step closure → `FusedOptimizer.AdamWUpdateSimd`.

**Exact cause:** the per-param loop guarded only against an empty *gradient* buffer (`gradArrays[p].Length == 0`), not an empty *param backing*. UnifiedMultimodal's unexercised/lazy params (the unused-modality encoders for a single-`[768]` input — e.g. a 295M-elem embedding) report their declared length but their live backing array is **empty (length 0)** until materialization. The plan still pre-allocates their grad/m/v buffers, so `fixed (float* pParam = paramArrays[p])` over the empty backing yields a **null pointer**, and the SIMD kernel deref's it at the full declared length → NRE → silent fallback to the eager tape for the whole model.

**Fix (PR #712):** skip params with no materialized backing (`paramArrays[p].Length == 0`), mirroring the eager path. Verified end-to-end:
- Fused path now **engages** (no fallback warning).
- Per-step managed allocation: **2318 → 21 MB/iter (111×)** — eliminates the gen2-GC churn / OOM pressure.
- Training still converges (loss 4.93 → ~0).
- Simple single-path models (FFN) unaffected — they always materialize, so the guard never fires.

## Honest status on the 120s timeout
This fixes the **allocation/OOM** axis decisively and puts the model on the industry-standard fused path. But wall-time stays ~1 s/step (≈250s for the 250-step MoreData) because the workload is **memory-bandwidth-bound 540M-param compute at batch=1**, not allocation-bound. So the HeavyTimeout tag stays for now — but the remaining gap is a *compute*-reduction lever (e.g. not computing unexercised-modality forward/backward), not the fused-path bug, which is now fixed.

**Delivery:** Tensors PR #712 → merge + release → bump `AiDotNet.Tensors` PackageVersion here. The release is the gated step.

**Comment by ooples** [comment]
## Compute-lever profiling (PerfView attempted; dotnet-trace used for parseable stacks)

After the fused-path fix engages, the per-step decomposition is:
- ForwardForTraining: **12 ms**
- Full Train step: **~955 ms**  → backward+optimizer ≈ **942 ms** (75× the forward)

**The optimizer is NOT the bottleneck.** Instrumenting the plan shows it processes only **64 params / 29M elems** (the exercised text path) and correctly **skips 38 params / 501M elems** (the unused-modality encoders — the #1738 guard working).

**The bottleneck is weight *initialization* running during training.** The call path (from sampled stacks):
`TryAutoEnableWeightStreaming → ConfigureWeightLifetime → RegisterTrainableTensorsWithWeightRegistry → DenseLayer.GetTrainableParameters → EnsureInitialized → InitializeParameters → XavierFill`

The model auto-enables weight streaming (540M > 500M threshold) and the streaming registration walk materializes + Xavier-initializes **all** params — including the ~500M unused-modality ones — on the training path (~0.86 s/iter of pure XavierFill, ~44% of the step in the parallel fill executor).

**But the obvious fix backfired.** Adding a "don't stream when the model fits in RAM" guard made the step ~2× *slower* (~1.9 s/iter): without streaming's lazy handling, all 540M params fully materialize and the optimizer then processes 540M instead of 29M. So streaming is paradoxically what keeps the unused params lazy/skipped. Reverted.

**Conclusion:** this is a delicate lazy-materialization × fused-plan × streaming interaction. The real fix is to stop the streaming registration from re-materializing/re-initializing **unused lazy** layers each step (init once and persist, or skip shape-unresolved lazy layers in the registration walk) — *without* losing the lazy-skip that keeps the optimizer at 29M. That needs focused work on a quiet box (current measurements are noisy, 0.8–2.2 s/iter). The fused-path NRE fix (Tensors PR #712) stands on its own — it's the larger, broadly-applicable correctness+allocation win.

**Comment by ooples** [comment]
## SOLVED — UnifiedMultimodal under the 120s gate (PerfView found it)

Using **PerfView** (kernel CPU sampling, admin — which resolves the native frames dotnet-trace shows only as "unmanaged") on a fused-path training step, the hot path was unambiguous:

`Train → fused step → CompiledTrainingPlan.StepEager → **ClipGradientsGlobalL2 = 44.4%**`

The actual work was tiny by comparison — GEMM ~5%, GEMV ~3.8%, the fused AdamW kernel just **1.4%**. The global-L2 gradient clip dominated.

**Root cause:** the clip scanned EVERY gradient buffer the plan holds, including the ~500M-element zero gradients of the unused-modality encoders (a single-`[768]` input only exercises the text path). Those gradients are zero — no contribution to the norm, scaling them is a no-op — and for the unmaterialized ones the `?? GetDataArray()` fallback even **materialized a full zero array** and scalar-scanned it twice (norm + scale) every step.

**Fix (Tensors PR #712):** `ClipGradientsGlobalL2` now skips a gradient whose owning parameter has no materialized weight backing (aligned with `_parameters`), mirroring what the fused optimizer already does. Math unchanged (skipped gradients are exactly zero).

**Result (measured, 540M-param model, single-modality input):**
| | before | after |
|---|---|---|
| steady train step | ~900 ms | **~350 ms (~2.5×)** |
| 250-step MoreData | ~225 s | **~87 s** ✅ under the 120s gate |
| loss trajectory | 4.93 → ~0 | 4.93 → ~0 (identical) |

So UnifiedMultimodal is no longer inherently HeavyTimeout — it was a fused-plan inefficiency.

### Deploy sequence
1. Merge Tensors PR #712 (the 6 fused-path NRE guards + this clip perf fix) → release new `AiDotNet.Tensors`.
2. Bump `AiDotNet.Tensors` PackageVersion in AiDotNet.
3. Drop UnifiedMultimodal's `HeavyTimeout` trait (added in PR #1739) — it now fits the default gate.

Steps 1–2 are gated on the package release.
